src/bin/fd_readdir.rs

```diff
@@ -0,0 +1,178 @@
+use libc;
+use misc_tests::open_scratch_directory;
+use misc_tests::wasi_wrappers::{wasi_fd_filestat_get, wasi_fd_readdir, wasi_path_open};
+use std::{cmp::min, env, mem, process, slice, str};
+use wasi::wasi_unstable;
+
+const BUF_LEN: usize = 256;
+
+#[derive(Debug)]
+struct DirEntry {
+    dirent: wasi_unstable::Dirent,
+    name: String,
+}
+
+// Manually reading the output from fd_readdir is tedious and repetitive,
+// so encapsulate it into an iterator
+struct ReadDir<'a> {
+    buf: &'a [u8],
+}
+
+impl<'a> ReadDir<'a> {
+    fn from_slice(buf: &'a [u8]) -> Self {
+        Self { buf }
+    }
+}
+
+impl<'a> Iterator for ReadDir<'a> {
+    type Item = DirEntry;
+
+    fn next(&mut self) -> Option<DirEntry> {
+        unsafe {
+            if self.buf.len() == 0 {
+                return None;
+            }
+
+            // Read the data
+            let dirent_ptr = self.buf.as_ptr() as *const wasi_unstable::Dirent;
+            let dirent = *dirent_ptr;
+            let name_ptr = dirent_ptr.offset(1) as *const u8;
+            // NOTE Linux syscall returns a NULL-terminated name, but WASI doesn't
+            let namelen = dirent.d_namlen as usize;
+            let slice = slice::from_raw_parts(name_ptr, namelen);
+            let name = str::from_utf8(slice).expect("invalid utf8").to_owned();
+
+            // Update the internal state
+            let delta = mem::size_of_val(&dirent) + namelen;
+            self.buf = &self.buf[delta..];
+
+            DirEntry { dirent, name }.into()
+        }
+    }
+}
+
+fn exec_fd_readdir(fd: wasi_unstable::Fd, cookie: wasi_unstable::DirCookie) -> Vec<DirEntry> {
+    let mut buf: [u8; BUF_LEN] = [0; BUF_LEN];
+    let mut bufused = 0;
+    let status = wasi_fd_readdir(fd, &mut buf, BUF_LEN, cookie, &mut bufused);
+    assert_eq!(status, wasi_unstable::ESUCCESS, "fd_readdir");
+
+    let sl = unsafe { slice::from_raw_parts(buf.as_ptr(), min(BUF_LEN, bufused)) };
+    let dirs: Vec<_> = ReadDir::from_slice(sl).collect();
+    println!("{:?}", dirs);
```

Comment:
Would you mind removing all `println!`s from the test please? :-)

Reply:
done.